### ISSUE TITLE
Distributed Tables missing records. "Exception: Too large string size." in logs.

### ISSUE THREAD

**Issue by entropical** [ClickHouse/ClickHouse#21894]
Setup:
         <shard>
              <weight>1</weight>
              <internal_replication>false</internal_replication>
              <replica>
                  <host><server1domainname></host>
                  <port>9000</port>
              </replica>
              <replica>
                  <host><server2domainname></host>
                  <port>9000</port>
              </replica>
          </shard>

Our code writes to a Distributed table on either server, and then clickhouse writes the record to the underlying real tables on both servers. 

No zookeeper.

This has been working for a few years.

Incident.
To get multivolume support, I upgraded from 18.14.19 to 20.8.12.2. I had a number of problems, and ran out of time to debug/fix, so I rolled back to 18.14.19. My problem probably started then.

Now we have discovered that some records are not getting replicated to one of the servers. Or maybe replication is working from server1 -> server2, but not the other way. I'm not 100% sure. When reads randomly choose a server, they get different results.

I have these error messages constantly on one server:
2021.03.19 02:35:13.294443 [ 39 ] {} <Error> <table>.Distributed.DirectoryMonitor: Poco::Exception. Code: 1000, e.code() = 0, e.displayText() = Exception: Too large string size., e.what() = Exception

I also noticed a small number of these errors on the other server, but they are not constant. Only a handful from a few hours ago.
2021.03.18 19:22:36.647630 [ 38 ] {} <Error> <table>.Distributed.DirectoryMonitor: Code: 252, e.displayText() = DB::Exception: Received from <serverdomainname>:9000, <ipaddress> DB::Exception: Too many parts (304). Merges are processing significantly slower than inserts..

I did some research into the first error, but I don't really understand which directory, or which string it is referring to.

Thanks.

**Comment by den-crane** [comment]
You don't have actual Replication. 
Distributed table propagates inserts into 2 CH nodes. Sometimes inserts can fail into one replica and succeed into another.
There is no surprise that Replicas has different data. It's documented behavior for this setup.

 >Poco::Exception. Code: 1000, e.code() = 0, e.displayText() = Exception: Too large string size

You should stop insertion and check what is going in Distributed table catalog. 
Probably drop Distributed table. Remove all remains (Distributed) folder and create Distributed table back.

> Too many parts (304). Merges are

Need to check what is going on with `<serverdomainname>:9000` in this server logs. And in `select * from system.merges`

**Comment by den-crane** [comment]
>I did some research into the first error, but I don't really understand which directory, or which string it is referring to.

It's about `/var/lib/clickhouse/data/db/distributedtablename/.....`

**Comment by entropical** [comment]
> You don't have actual Replication.
> It's documented behavior for this setup.

Yeah, I did know this, but up until very recently it has been sufficient for our purpose.

> It's about /var/lib/clickhouse/data/db/distributedtablename/.....

Right! I noticed this directory when I had the failed upgrade.

I have two directories here, but I think I should only have one. One has IP, one has domainname. I'm pretty sure this is related to the version change.

/var/lib/clickhouse/data/noc/distributedtablename/
default@111%2E222%2E333%2E444:9000/
default@domainname%2Etld%2Etld%2Etld%2Ecom:9000/

> Probably drop Distributed table. Remove all remains of Distributed folder (/var/lib/clickhouse/data/db/distributedtablename/) and create Distributed table back.

Yeah, this sounds like it might help. This is safe, right, because the Distributed table doesn't actually contain data? Trying not to lose any (more...) data.

**Comment by den-crane** [comment]
>This is safe, right, because the Distributed table doesn't actually contain data? 

Yes. It is safe. It contains only not propagated inserts. That's why I suggest to stop insertion.
Check that it stops to propagate. 
Check the size of these subdirectories: `du -sh /var/lib/clickhouse/data/noc/distributedtablename/*`
If it's huge, check how many .bin files inside. Check the server logs.

**Comment by entropical** [comment]
On one server, both of these directories are very small.
On the other server, the one with the IP address in as the name is small. The one with the domain name appears to be so huge I can't even ls or du it.

> Remove all remains of Distributed folder

Hypothetically, if there are orphaned bin files here, can I reintroduce them after I recreate the Distributed table, and it is working? I'm not _too_ concerned if not, just working out my options.

Thanks for your excellent help by the way.

**Comment by den-crane** [comment]
>Hypothetically, if there are orphaned bin files here, 
>can I reintroduce them after I recreate the Distributed table, and it is working?
> I'm not too concerned if not, just working out my options.

If you sure that both directories are the same destination you can move files from the one to anther.
And yes, you can move files / rename (move) the whole  directory to somewhere to /var/lib/clickhouse
recreate the distr. table and move .bin files one by one or  1000 by 1000 ( batches ).

**Comment by entropical** [comment]
I stopped my writers, and am now dropping the replicatedtable. It has taken about 10 minutes so far...

**Comment by entropical** [comment]
The drop was taking forever. I think it was attempting to catch up on the pending propogations first.

I moved the directory to a safe location, and the drop completed a few seconds later. I then recreated the distributed table, and started my writers. For the last hour it looks like the propagations are working fine.

I'll now try to reparent my orphan bin files.

Thanks again for your help @den-crane.
